src/app-logic/browser-connection.js

```diff
@@ -72,42 +66,52 @@ class BrowserConnectionImpl implements BrowserConnection {
       webChannelSupportsGetProfileAndSymbolication;
   }
 
-  async establishConnectionViaFrameScriptIfNeeded(options: {|
+  // Only called when we must obtain the profile from the browser, i.e. if we
+  // cannot proceed without a connection to the browser. This method falls back
+  // to the frame script API (window.geckoProfilerPromise) if this browser has
+  // an old version of the WebChannel.
+  async _getConnectionViaFrameScript(): Promise<$GeckoProfiler> {
+    if (!this._geckoProfiler) {
+      this._geckoProfiler = await window.geckoProfilerPromise;
+    }
+    return this._geckoProfiler;
+  }
+
+  async getProfile(options: {|
     onThirtySecondTimeout: () => void,
-  |}) {
-    // If the profiler was opened at the /from-browser URL, and this is not a restored
-    // Firefox session, then we're likely to have a geckoProfiler object. Wait forever.
+  |}): Promise<ArrayBuffer | MixedObject> {
     const timeoutId = setTimeout(options.onThirtySecondTimeout, 30000);
-    this._geckoProfiler = await window.geckoProfilerPromise;
-    clearTimeout(timeoutId);
-  }
 
-  async getProfile(): Promise<ArrayBuffer | MixedObject> {
     // On Firefox 96 and above, we can get the profile from the WebChannel.
     if (this._webChannelSupportsGetProfileAndSymbolication) {
-      return getProfileViaWebChannel();
+      const profile = await getProfileViaWebChannel();
+      clearTimeout(timeoutId);
+      return profile;
     }
+
     // For older versions, fall back to the geckoProfiler frame script API.
     // This fallback can be removed once the oldest supported Firefox ESR version is 96 or newer.
-    if (this._geckoProfiler) {
-      return this._geckoProfiler.getProfile();
-    }
-
-    throw new Error(
-      'Cannot obtain a profile: have neither WebChannel nor a GeckoProfiler object'
-    );
+    const geckoProfiler = await this._getConnectionViaFrameScript();
+    const profile = await geckoProfiler.getProfile();
+    clearTimeout(timeoutId);
+    return profile;
   }
 
-  getSymbolTable(
+  async getSymbolTable(
     debugName: string,
     breakpadId: string
   ): Promise<SymbolTableAsTuple> {
     // On Firefox 96 and above, we can get the symbol table from the WebChannel.
     if (this._webChannelSupportsGetProfileAndSymbolication) {
       return getSymbolTableViaWebChannel(debugName, breakpadId);
     }
+
     // For older versions, fall back to the geckoProfiler frame script API.
     // This fallback can be removed once the oldest supported Firefox ESR version is 96 or newer.
+    // Note that we use this._geckoProfiler directly instead of
+    // _getConnectionViaFrameScript so that we're not waiting forever when the
+    // user opens an unsymbolicated profile with a Firefox that doesn't support
+    // the WebChannel.
```

Comment:
Let's also add the following:

```js
// This means that we'll only ask the frame script for symbol tables if we also 
// obtained the profile from the frame script, which is fine because that's the
// only case in which the frame script would be loaded in this tab anyway.
```

Reply:
Actually, maybe we don't need this comment. Not many people will read this code and have the question "but what about symbolicating profiles from files with the old webchannel?"... it's more just for my own peace of mind.